### DISCUSSION TITLE
Knowing current firmware version of connected USB devices

### DISCUSSION THREAD

**Discussion by vijay-bhakta** [fwupd/fwupd#5736] [General]
The first step of any upgrade mechansim is knowing the current version itself.

1. In general, can LVFS detect and provide current firmware version of any plugged-in USB device if following one of supported LVFS mechanism regardless of the fact the device vendor associated and having account with LVFS or not?
2. How does LVFS fetch the current firmware version for a USB connected device given the command `fwupdmgr get-devices` ?
3. If I am a vendor making a new USB device, what standard mechanism recommended to me by LVFS that i do not  depend on any plugin to let LVFS retrieve my current firmware version? Is it a USB spec or LVFS defined standard to follow?

I came across USB standard DFU 1.1 ( https://www.usb.org/document-library/device-firmware-upgrade-11-new-version-31-aug-2004 ) which apparently specifies mechanism to push new firmware to USB-device, but has no mention regarding retrieving current firmware

**Comment by hughsie** [comment]
> can LVFS detect and provide current firmware version of any plugged-in USB device

Yes and no -- in the most basic case we can read the BCD bcdVersion from the USB interface -- this is enough for some protocols that only want a AA.BB version number.

> regardless of the fact the device vendor associated and having account with LVFS or not

Enumeration of the device is unrelated to if firmware updates are being shipped.

> How does LVFS fetch the current firmware version for a USB

Depends on the device protocol. For USB devices we normally read bcdVersion and then override this with something more helpful if specified as part of the protocol. e.g.BCD `1.2` could turn into `1.2.20210421` using a vendor-specific HID or control transfer request.

> If I am a vendor making a new USB device

If you're doing this from scratch using DFU means you don't need to write any code or configuration. Some DFU devices need quirk files for bugs in the implementation, but in the general case they just work. If you can stomach a AA.BB version the bcdVersion is usually *good enough* for most hardware.

**Reply by vijay-bhakta** [reply]
> For USB devices we normally read bcdVersion

I read about it in DFU 1.1 pdf section 4.2 .  My understanding is that this part comes after the host initiatiated DFU mode , and device becomes non-operational.
Are you saying LVFS does this everytime on command `fwupdmgr get-devices` to current firmware version of connected USB device?

**Reply by vijay-bhakta** [reply]
> AA.BB version 

Its has 2 byte space for `major.minor` version representation.
does it mean one byte for major number and other byte minor?

**Reply by hughsie** [reply]
>  My understanding is that this part comes after the host initiatiated DFU mode

Nah, fwupd reads the bcdVersion for *all* USB devices, regardless of protocol they support.

> does it mean one byte for major number and other byte minor

Almost. It's BCD, i.e. https://en.wikipedia.org/wiki/Binary-coded_decimal

**Reply by vijay-bhakta** [reply]
For USB device manufacturer's to support standard mechanism for upgrade and hence compatible on LVFS without plugins

1. **Support bcdDevice in Device Descriptor**:  the host can request [USB_DEVICE_DESCRIPTOR](https://www.keil.com/pack/doc/mw/USB/html/group__usbh__data__types.html#struct_u_s_b___d_e_v_i_c_e___d_e_s_c_r_i_p_t_o_r) at any time to know current version, which is similar to one defined in [DFU 1.1 PDF ](https://www.usb.org/sites/default/files/DFU_1.1.pdf) Section 4.2.1.  And LVFS command `fwupdmgr get-devices` use the same .
2. **Support DFU protocol:** To push the new firmware to device where host initiates DFU

correct?

We make a hardware with Linux where we connect USB cameras. I am just looking for basic specification that we can ask our USB device partners to adhere to so that they can be LVFS compatible.

**Reply by vijay-bhakta** [reply]
any comment on above statement ?

**Reply by hughsie** [reply]
Using any of the existing protocols (especially DFU) would be great, but we already support dozens of vendor specific protocols in plugins, so it's not the end of the world to add one more. It depends on how expensive it would be to change the device firmware update mechanism, Vs mandating contributing to the fwupd project.

**Comment by vijay-bhakta** [comment]
This brings me to question that why do these manufacturers implemented something of there own and now need to write plugins in LVFS?

It is interesting to see USB device ecosystem that , even though , USB forum has well defined specifications (e.g. DFU 1.1 since 2004) what did really happen.

**Reply by vijay-bhakta** [reply]
In your knowledge, how many vendors do you think follow  _bcdDevice_ in Device Descriptor?
The first focus for us is to inform the current version of ANY connected USB camera device to end-user. 

Currently, I see many of the Camera devices are not even listed when called `fwupdmgr get-devices`

**Reply by hughsie** [reply]
> This brings me to question that why do these manufacturers implemented something of there own

That's a very good question, and one with a mostly political answer. Most of the time it's engineering for job security or to support legacy hardware. DFU hasn't been updated in a while, and the USB consortium isn't helping there.

> Currently, I see many of the Camera devices are not even listed when called fwupdmgr get-devices

We only show things that are updatable with any supported protocol. Have a look with `lsusb -vvvv` to get *everything*.
